roles/irods/tasks/icat.yml

```diff
@@ -1,29 +1,167 @@
+    mode: '0600'
   register: create_unattended_installation_json
+  become: true
+
+# - name: Check if {{ irods_db_name }} database is available
+#  postgresql_ping:
+#    db: "{{ irods_db_name }}"
+#  register: check_database
+#  become_user: postgres
+#  become: true
+#
+- name: Create a new database "{{ irods_db_name }}" if missing
+  postgresql_db:
+    name: "{{ irods_db_name }}"
+    state: present
+#  when: check_database.is_available is false
+  register: db_created
+  become_user: postgres
+  become: true
 
-- name: run_irods_installation_script
-  become: True
-  shell:  python /var/lib/irods/scripts/setup_irods.py --json_configuration_file=/root/unattended_install.json
-  when: create_unattended_installation_json.changed
+- name: Creating irods sql user and granting privileges
+  postgresql_user:
+    name: "{{ irods_db_user }}"
+    password: "{{ irods_db_pwd }}"
+    db: "{{ irods_db_name }}"
+    priv: ALL
+    state: present
+  become_user: postgres
+  become: true
+
+- name: Collect list of tables from existing {{ irods_db_name }}
+  ansible.builtin.command: psql -d ICAT -c '\dt'
+  changed_when: false
+  register: database_tables
+  become_user: postgres
+  become: true
+
+- name: Install iRODS, if {{ irods_db_name }} is empty (no tables exist)
+  ansible.builtin.command: sudo python /var/lib/irods/scripts/setup_irods.py --json_configuration_file=/root/unattended_install.json
+  when: '"No relations found." in database_tables.stdout'
   run_once: true
   async: 120
-  poll: 1  
+  poll: 1
+  become_user: '{{ irods_service_account }}'
+  become: true
+
+- name: Edit /etc/irods/core.re to enforce the use of the certificates
+  become: true
+  lineinfile:
+    name: /etc/irods/core.re
+    search_string: 'acPreConnect(*OUT) { *OUT="CS_NEG_DONT_CARE"; }'
+    line: 'acPreConnect(*OUT) { *OUT="CS_NEG_REQUIRE"; }'
+
+- name: Edit /etc/irods/core.re to set number of threads to 4
+  become: true
+  lineinfile:
+    name: /etc/irods/core.re
+    search_string: 'acSetNumThreads {msiSetNumThreads("default","default","default"); }'
+    line: 'acSetNumThreads {msiSetNumThreads("default","4","default"); }'
+
+- name: Increase systcl limits to allow longer transfers
+  copy:
+    src: '{{ role_path }}/files/irods_fix.conf'
+    dest: /etc/sysctl.d/
+    mode: 0644
+  become: true
+
+- name: Fix the s3 authentication by creating dummy s3auth file
+  ansible.builtin.copy:
+    content: |
+      foo
+      bar
+    dest: /etc/irods/.s3auth
+    force: true
+    mode: 0644
+    owner: '{{ irods_service_account }}'
+    group: '{{ irods_service_account }}'
+  become_user: '{{ irods_service_account }}'
+  become: true
+
+- name: Enable and iRODS service
+  sysvinit:
+    name: irods
+    enabled: true
+  become: true
+
+# Start/stop/restart service call fails if with error if already in that state
+# so it's best if done via user call (as this is what the service does anyways)
+- name: Restart iRODS server manually via {{ irods_service_account }}
```

Comment:
Services should only be restarted if necessary. E.g. when some configuration has changed or software got updated. This should be moved to a `handler`.

Reply:
:heavy_check_mark:  The new commit has (tested successfully) handlers.

Why the system service was not used: (the explanation for the rest):

Sysvinit service cannot be used, as ansible calls support only `start`, `stop`, `restarted` and `reloaded`:
  - the `start` does not work - even when irods is shut down, it does not start it up (repeatably) and subsequent Davrods role fails
  - the `restarted` repeatably fails upon new installation when service is still `stopped` (see output below, but in short: service uses lock file `/var/lock/subsys/irods` that cannot remove it, as it does not exists). I am certain this is a bug in a `/etc/init.d/irods`
- since the `stop` and `reloaded` (which does not help, nor is implemented in irods service definition file) this means the native sysvinit service call cannot be used.

My implementation avoids use of service call and directly restarts the server in the same way as the service itself would (I looked into the `/etc/init.d/irods`, it is basically just `/var/lib/irods/irodsctl` script). Direct call does not fail the role, and we do not need to add into ansible the wrappers to catch errors. It also only fails when there is an actual error with irods server startup.

Problem with system service `restarted` call:
```
TASK [irods : Enable and iRODS service] *********************************************************************************************************************
fatal: [tunnel+irods-test]: FAILED! => changed=false 
  msg: 'Failed to restart service: irods'
  rc: 1
  stderr: |-
    [1;33mNo iRODS servers running.[0m
    rm: cannot remove ‘/var/lock/subsys/irods’: No such file or directory
  stderr_lines: <omitted>
  stdout: ''
  stdout_lines: <omitted>

```

This bug can be also resolved by changing the `/etc/init.d/irods` line:
`    if [ "$DETECTEDOS" = "RedHatCompatible" -o "$DETECTEDOS" = "SuSE" ]; then`
into
`    if [ "$DETECTEDOS" = "RedHatCompatible" -o "$DETECTEDOS" = "SuSE" ] && [ -e /var/lock/subsys/irods ]; then`

Reply:
Bug confirmed, going to be fixed in upcoming versions: https://github.com/irods/irods/issues/6053